### PR TITLE
CHANGE triangulation_earclip

### PR DESCRIPTION
**Pull request by petrasvestartas**
The change of `compas\src\compas\geometry\triangulation_earclip.py ` code to a new one.
The issue with the current algorithm is described in the following forum thread:
[COMPAS FORUM](https://forum.compas-framework.org/t/mesh-face-grouping/719/2)

The function name stays the same:

```python

if __name__ == "__main__":
    points = [
        [377.952174, -3452, 1500.484283],
        [369.348527, -3265, 1488.197003],
        [459.973604, -3252, 1617.623026],
        [595.337643, -3295, 1810.942908],
        [718.083001, -3144, 1986.241446],
        [458.252875, -3120, 1615.16557],
        [336.081094, -3052, 1440.686184],
        [277.576297, -3227, 1357.132676],
        [219.071501, -3154, 1273.579167],
        [74.530239, -3315, 1067.152852],
        [69.368051, -3436, 1059.780484],
        [86.575344, -3695, 1084.355045],
        [-2.329003, -3686, 957.386478],
        [-51.083001, -3865, 887.758554],
        [63.05871, -4035, 1050.769811],
        [136.476494, -3688, 1155.621273],
        [192.686985, -3362, 1235.898173],
        [332.639635, -3414, 1435.771272],
        [278.149874, -3619, 1357.951828],
        [186.377644, -3672, 1226.887501],
        [189.245526, -3981, 1230.983261],
        [339.522553, -4195, 1445.601096],
        [441.619158, -3998, 1591.41016],
        [448.502075, -3634, 1601.239985],
        [477.180897, -3578, 1642.197587],
        [505.286143, -3908, 1682.336037],
        [642.370911, -3534, 1878.113376],
        [422.117559, -3301, 1563.558991],
        [332.639635, -3824, 1435.771272],
    ]

    polygon = Polygon(points)
    triangles = earclip_polygon(polygon)
    print(triangles)
```

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner. - replacement of existing code.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas/geometry/triangulation_earclip.py [hunk 1]
```diff
@@ -1,8 +1,248 @@
-from compas.geometry import is_ccw_xy, is_point_in_triangle_xy
+class Ear:
```
Comment (gonzalocasas) [inline src/compas/geometry/triangulation_earclip.py]: python 2.x old-style classes compatibility:

```suggestion
class Ear(object):
```

File: src/compas/geometry/triangulation_earclip.py [hunk 2]
```diff
@@ -1,8 +1,248 @@
...
+
+        Returns
+        -------
+        list: List of vertex indices forming the Ear triangle.
+
+        """
+        return [self.prew, self.index, self.next]
+
+
+class Earcut:
```
Comment (gonzalocasas) [inline src/compas/geometry/triangulation_earclip.py]: Same here 
```suggestion
class Earcut(object):
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -55,6 +55,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Changed `compas.datastructures.Mesh` to take additional `**kwargs`, instead of only `name=None` specifically.
 * Moved registration of `ping` and `remote_shutdown` of the RPC server to `compas.rpc.Server.__init__()`.
 * Moved `FileWatcherService` to `compas.rpc.services.watcher` so it can be reused.
+* Changed `compas.geometry.earclip_polygon` algorithm because there current one does not work in several cases.
```
Comment (gonzalocasas) [inline CHANGELOG.md]:
```suggestion
* Changed `compas.geometry.earclip_polygon` algorithm because the current one does not work in several cases.
```

**Review by gonzalocasas — COMMENTED** [review]
The code looks good, I didn't check the math though. But it is missing tests, and since you mention that the previous algo failed in some cases, I would like to see those failing cases added as unit tests + some non-edge cases.

File: src/compas/geometry/triangulation_earclip.py [hunk 4]
```diff
@@ -1,8 +1,248 @@
...
+        coords : list
+            Coordinates of the vertex of the Ear triangle.
+        next : int
+            Index of the next vertex of the Ear triangle.
+        prew : int
+            Index of the previous vertex of the Ear triangle.
+        neighbour_coords : list
+            Coordinates of the next and previous vertices of the Ear triangle.
+
+        """
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]: add to class docstring

File: src/compas/geometry/triangulation_earclip.py [hunk 5]
```diff
@@ -1,8 +1,248 @@
...
+        if d[0] * d[1] >= 0 and d[2] * d[1] >= 0 and d[0] * d[2] >= 0:
+            return True
+        return False
+
+    def is_ear_point(self, p):
+        """Check if a given point is one of the vertices of the Ear triangle.
+
+        Returns
+        -------
+        bool : True, if the point is a vertex of the Ear triangle, False otherwise.
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]:
```python
bool
    True, if ...
```

File: src/compas/geometry/triangulation_earclip.py [hunk 6]
```diff
@@ -1,8 +1,248 @@
...
+        if p == self.coords or p in self.neighbour_coords:
+            return True
+        return False
+
+    def validate(self, points, indexes, ears):
+        """Validate if the Ear triangle is a valid Ear by checking its convexity and that no points lie inside.
+
+        Returns
+        -------
+        bool: True if the Ear triangle is valid, False otherwise.
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]: same

File: src/compas/geometry/triangulation_earclip.py [hunk 7]
```diff
@@ -1,8 +1,248 @@
...
+                    return False
+            return True
+        return False
+
+    def is_convex(self):
+        """Check if the Ear triangle is convex.
+
+        Returns
+        -------
+        bool: True if the Ear triangle is convex, False otherwise.
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]: same

File: src/compas/geometry/triangulation_earclip.py [hunk 8]
```diff
@@ -1,8 +1,248 @@
...
+        if ab[0] * bc[1] - ab[1] * bc[0] <= 0:
+            return False
+        return True
+
+    def get_triangle(self):
+        """Get the indices of the vertices forming the Ear triangle.
+
+        Returns
+        -------
+        list: List of vertex indices forming the Ear triangle.
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]: same

File: src/compas/geometry/triangulation_earclip.py [hunk 9]
```diff
@@ -1,8 +1,248 @@
...
+        ears : list
+            List of Ear objects representing the Ears of the polygon.
+        neighbours : list
+            List of indices of the neighbouring vertices.
+        triangles : list
+            List of triangles forming the triangulation of the polygon.
+        length : int
+            Number of vertices of the polygon.
+
+        """
```
Comment (tomvanmele) [inline src/compas/geometry/triangulation_earclip.py]: add to class docstring

**Comment by petrasvestartas** [comment]
Thank you for a review, the changes have been made, including pytest, and pushed once more.

**Comment by tomvanmele** [comment]
can you merge main back in please...

**Comment by petrasvestartas** [comment]
`earclip_polygon` method is merged back.

File: CHANGELOG.md [hunk 10]
```diff
@@ -82,6 +82,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Moved `compas.numerical.matrices` to `compas.topology.matrices`.
 * Moved `compas.numerical.linalg` to `compas.geometry.linalg`.
 * Changed `watchdog` dependency to be only required for platforms other than `emscripten`.
+* Changed `compas.geometry.earclip_polygon` algorithm because the current one does not work in several cases.
+* Changed the docstrings in `compas.geometry.earclip_polygon` are corrected. The orientation to the XY Frame has been removed since it is handled by the `polygon.to_vertices_and_faces` method. Additionally, the test_polygon.py file has been modified with a new test for the ear_clip called `test_polygon_to_vertices_and_faces`. The `object` has been added to all `Ear` classes.
+* Changed the docstrings in `compas.geometry.earclip_polygon` is merged back, meaning the polygon is transformed to XY frame.
```
Comment (gonzalocasas) [inline CHANGELOG.md]: Don't need to add tests and docstring stuff to the changelog, the changelog is only for users of the library that need to understand what has changed

File: src/compas/geometry/triangulation_earclip.py [hunk 11]
```diff
@@ -1,8 +1,251 @@
...
+                self.add_ear(prew_ear_new)
+                continue
+            if next_ear_new.validate(self.vertices, indexes, self.ears) and next_ear_new.index not in self.neighbours:
+                self.add_ear(next_ear_new)
+                continue
+
+        return self.triangles
 
 
 def earclip_polygon(polygon):
```
Comment (gonzalocasas) [inline src/compas/geometry/triangulation_earclip.py:252]: The unit test added does not explicitly test the earclip algorithm, and actually I find it hard to see the link from the code. I would add an additiona `tests/compas/geometry/test_triangulation_earclip.py` that tests directly the relevant methods of this module.

Comment (petrasvestartas) [reply src/compas/geometry/triangulation_earclip.py:252]: I have just made a new push to a new file: `tests/compas/geometry/test_triangulation_earclip.py` for testing.

I added one small detail to handle simple triangles.

File: CHANGELOG.md [hunk 12]
```diff
@@ -82,6 +82,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Moved `compas.numerical.matrices` to `compas.topology.matrices`.
 * Moved `compas.numerical.linalg` to `compas.geometry.linalg`.
 * Changed `watchdog` dependency to be only required for platforms other than `emscripten`.
+* Changed `compas.geometry.earclip_polygon` algorithm because the current one does handle several cases.
```
Comment (gonzalocasas) [inline CHANGELOG.md]: Tiny typo ;)
```suggestion
* Changed `compas.geometry.earclip_polygon` algorithm because the current one does not handle several cases.
```

**Review by gonzalocasas — APPROVED** [review]
looks good to me! Besides the little typo, there's some linting required (the build is failing because of that). Once the build is green, I think this is good to go!

**Comment by tomvanmele** [comment]
@petrasvestartas run `invoke check` to find all such errors locally...

**Comment by tomvanmele** [comment]
to apply `black` to the src folder and fix things automatically `black src`
